### ISSUE TITLE
[Bug] casting deeply nested paths does not work with `updateOne` but works with `create`

### ISSUE THREAD

**Issue by AbdelrahmanHafez** [Automattic/mongoose#10200]
Passing deeply nested paths on `updateOne` should convert the path to a deeply structured object, which works with new Model(...) and Model.create(...), but does not work with `Model.updateOne(...)` and `Model.bulkWrite(...)`.

Repro script:
### 10200.js
```js
describe('converts dot separated paths to nested structure', () => {
  it('works with new Model(...)', () => { // passes in 5.12.7
    const Payment = getPaymentModel();
    const paymentPOJO = getPaymentWithDotSeparatedPaths();
    const payment = new Payment(paymentPOJO);
    assertDocumentStructure(payment.toObject());
  });
  it('works with Model.create(...)', () => { // passes in 5.12.7
    return co(function*() {
      const Payment = getPaymentModel();
      const paymentPOJO = getPaymentWithDotSeparatedPaths();
      const payment = yield Payment.create(paymentPOJO);
      assertDocumentStructure(payment);
    });
  });
  it('works with Model.updateOne(...)', () => { // fails in 5.12.7
    return co(function*() {
      const User = getPaymentModel();
      const userPOJO = getPaymentWithDotSeparatedPaths();

      const emptyUser = yield User.create({});
      yield User.updateOne({ _id: emptyUser._id }, userPOJO);
      const user = yield User.findOne({ _id: emptyUser._id }).lean();

      assertDocumentStructure(user);
    });
  });
  it('works with Model.bulkWrite(...)', () => { // fails in 5.12.7
    return co(function*() {
      const Payment = getPaymentModel();
      const paymentPOJO = getPaymentWithDotSeparatedPaths();

      const emptyPayment = yield Payment.create({});

      yield Payment.bulkWrite([
        { updateOne: { filter: { _id: emptyPayment._id }, update: paymentPOJO } }
      ]);
      const payment = yield Payment.findOne({ _id: emptyPayment._id }).lean();

      assertDocumentStructure(payment);
    });
  });


  function getPaymentModel() {
    const paymentSchema = new Schema({
      paymentFor: String,
      externalServiceResponse: {
        id: String,
        resultDetails: {
          clearingInstituteName: String,
          transaction: {
            receipt: String,
            authorizationCode: String,
            acquirer: { settlementDate: String }
          },
          response: { acquirerCode: String, acquirerMessage: String },
          authorizationResponse: { stan: String },
          sourceOfFunds: { provided: { card: { issuer: String } } }
        }
      }
    });

    const Payment = db.model('Payment', paymentSchema);
    return Payment;
  }

  function getPaymentWithDotSeparatedPaths() {
    return {
      paymentFor: 'order',
      externalServiceResponse: {
        id: '1',
        resultDetails: {
          clearingInstituteName: 'Our local bank',
          'authorizationResponse.stan': '123456',
          'transaction.receipt': 'I am a transaction receipt',
          'transaction.authorizationCode': 'ABCDEF',
          'transaction.acquirer.settlementDate': 'February 2021',
          'sourceOfFunds.provided.card.issuer': 'Big bank corporation',
          nonExistentField: 'I should not be present'
        }
      }
    };
  }

  function assertDocumentStructure(payment) {
    assert.equal(payment.paymentFor, 'order');
    assert.equal(payment.externalServiceResponse.id, '1');
    assert.equal(payment.externalServiceResponse.resultDetails.clearingInstituteName, 'Our local bank');
    assert.deepEqual(
      payment.externalServiceResponse.resultDetails.authorizationResponse,
      { stan: '123456' }
    );
    assert.deepEqual(
      payment.externalServiceResponse.resultDetails.transaction,
      {
        receipt: 'I am a transaction receipt',
        authorizationCode: 'ABCDEF',
        acquirer: { settlementDate: 'February 2021' }
      }
    );
    assert.deepEqual(
      payment.externalServiceResponse.resultDetails.sourceOfFunds,
      { provided: { card: { issuer: 'Big bank corporation' } } }
    );
  }
});
```

### Output
```
  2 passing (377ms)                                                                                      
  2 failing                                                                                              
                                                                                                         
  1) Model                                                                                               
       converts dot separated paths to nested structure                                                  
         works with Model.updateOne(...):                                                                
     AssertionError [ERR_ASSERTION]: Expected values to be loosely deep-equal:                           
                                                                                                         
undefined                                                                                                
                                                                                                         
should loosely deep-equal                                                                                
                                                                                                         
{                                                                                                        
  stan: '123456'                                                                                         
}                                                                                                        
      at assertDocumentStructure (test\model.test.js:7296:14)                                            
      at mongoose\test\model.test.js:7231:9                     
      at Generator.next (<anonymous>)                                                                    
      at onFulfilled (node_modules\co\index.js:65:19)                                                    
      at runMicrotasks (<anonymous>)                                                                     
      at processTicksAndRejections (internal/process/task_queues.js:93:5)                                
                                                                                                         
  2) Model                                                                                               
       converts dot separated paths to nested structure                                                  
         works with Model.bulkWrite(...):                                                                
     AssertionError [ERR_ASSERTION]: Expected values to be loosely deep-equal:                           
                                                                                                         
undefined                                                                                                
                                                                                                         
should loosely deep-equal                                                                                
                                                                                                         
{                                                                                                        
  stan: '123456'                                                                                         
}                                                                                                        
      at assertDocumentStructure (test\model.test.js:7296:14)                                            
      at mongoose\test\model.test.js:7246:9                     
      at Generator.next (<anonymous>)                                                                    
      at onFulfilled (node_modules\co\index.js:65:19)                                                    
      at runMicrotasks (<anonymous>)                                                                     
      at processTicksAndRejections (internal/process/task_queues.js:93:5)                                
                                                                                                         
```

**Comment by AbdelrahmanHafez** [comment]
I am not sure if I understand correctly.
The way I understand the difference between both options can be demonstrated by the following example.
```js
const schema = new Schema({ a: { b: { c: Number, d: Number } } });
const document = { a: { b: { c: 1, d: 2 } } };

Example A
// $set: { 'a.b.c': 42 }
// the document should be
// { a: { b: { c: 42, d: 2 } } }

Example B
// $set: { a: { 'b.c': 42 } }
// the resulting document should be
// { a: { b: { c: 42 } } }
```
The way I phrase this in my head is:
Example A: change a.b.c to 42.
Example B: change a to an object that's { 'b.c': 24 }, which erases the value of `d`.

That makes sense to me, the issue I am facing is that in the schema the value is defined as nested objects, while the value reproduced in the test bypasses the shape defined in the schema.

```
Expected result for the document in MongoDB:
// { a: { b: { c: 42 } } }

Actual result for the document in MongoDB:
// { a: { 'b.c': 42 } }
```

**Comment by vkarpov15** [comment]
You make a good point, we'll fix this issue for v5.12.9. Since we don't support property names with '.' in schemas, we shouldn't allow updates with nested dotted properties to go through.

**Comment by AbdelrahmanHafez** [comment]
@vkarpov15 Thank you for fixing this. 👍
I attempted to fix it myself but walkUpdatePath was a bit intimidating and couldn't understand exactly where change should be.
